### PR TITLE
Write documentation with examples

### PR DESCRIPTION
**Pull request by mlange-42**

### CODE REVIEW

**Comment by mlange-42** [comment]
How can documentation be made foldable, like on docs.rs?

File: src/lib.rs [hunk 1]
```diff
@@ -1,3 +1,95 @@
...
+//! }
+//!
+//! for (_entity, number, string) in Query::<(&Entity, &u32, &String)>::new().iter(&world) {
+//!     println!("{}, {}", string, number);
+//! }
+//! ```
+//!
+//! ----
+//!
+//! A more complex example using custom components, resources, and functions as systems:
```
Comment (mlange-42) [inline src/lib.rs]: This might be a bit too much? However, I think is demonstrates well how to use it all in conjunction.

**Comment by adamreichold** [comment]
> Add doc-tests to CI

This is part of `cargo test` so it should already be included in the CI.

Please also note that I just impolitely force-pushed the main branch. Could you please rebase this to drop the stale CI commit?

File: src/lib.rs [hunk 2]
```diff
@@ -1,3 +1,95 @@
+//! A reasonably simple entity component system (ECS),
+//! with a design based on [hecs](https://github.com/Ralith/hecs).
+//!
+//! Compared to hecs, but it drops all optimisations which require additional API surface.
```
Comment (adamreichold) [inline src/lib.rs]: Please compare with the updated README. I think this is a bit too much and we should focus on the difference in approach, i.e. the simpler API.

**Comment by adamreichold** [comment]
> How can documentation be made foldable, like on docs.rs?

Could you post a screenshot of what you mean compared to what you get? Additionally, did you trying using `cargo +nightly doc`? I think docs.rs is not always running the stable version of rustdoc.

**Comment by adamreichold** [comment]
> > Add doc-tests to CI
> 
> This is part of `cargo test` so it should already be included in the CI.
> 
> Please also note that I just impolitely force-pushed the main branch. Could you please rebase this to drop the stale CI commit?

Just double-checked it and your doc-tests did run as part of the CI pipeline:
```
Doc-tests rs-ecs

running 9 tests
test src/lib.rs - (line 37) ... ok
test src/lib.rs - (line 13) ... ok
test src/world.rs - world::World::alloc (line 58) ... ok
test src/world.rs - world::World::contains (line 266) ... ok
test src/world.rs - world::World::free (line 94) ... ok
test src/world.rs - world::World::get (line 289) ... ok
test src/world.rs - world::World::get_mut (line 312) ... ok
test src/world.rs - world::World::insert (line 142) ... ok
test src/world.rs - world::World::remove (line 181) ... ok
```

File: src/lib.rs (hunk 1 above)
Comment (adamreichold) [reply src/lib.rs]: I think such "worked examples" are usually placed in the `examples` folder as a separate binary? Especially since this commented and can be read standalone, making it into `examples/motion.rs` or something similar should work better than browsing it via the docs?

**Comment by mlange-42** [comment]
> 
> 
> > How can documentation be made foldable, like on docs.rs?
> 
> Could you post a screenshot of what you mean compared to what you get? Additionally, did you trying using `cargo +nightly doc`? I think docs.rs is not always running the stable version of rustdoc.

At the top right of each doc page, there is `[-][src]`. In the docs for `World`, the `[-]` does not fold method docs. However, I just realized that is does in the docs for `Recources` (commited just now).

File: src/resources.rs [hunk 3]
```diff
@@ -4,16 +4,45 @@ use std::collections::hash_map::{Entry, HashMap};
 use std::hash::{BuildHasherDefault, Hasher};
 use std::ops::{Deref, DerefMut};
 
+/// A type map for holding resources.
+///
+/// In an ECS, resources are "global" singular objects that may be accessed by systems
```
Comment (adamreichold) [inline src/resources.rs]: I think the comparison with entities is not too helpful for a first entry. Maybe just "resources replace global variables and my be accessed by systems that know their type"?

File: src/lib.rs [hunk 4]
```diff
@@ -1,3 +1,95 @@
...
+//!     pub y: f32,
+//! }
+//!
+//! struct Velocity {
+//!     pub x: f32,
+//!     pub y: f32,
+//! }
+//!
+//! // A resource for the length of the time step
+//! struct Delta {
```
Comment (adamreichold) [inline src/lib.rs]: I think this would be a good tuple struct.

Comment (adamreichold) [reply src/lib.rs]: (The `dt` will still be visible in the let binding.)

File: src/lib.rs (hunk 2 above)
Comment (mlange-42) [reply src/lib.rs]: Done.

File: src/resources.rs [hunk 5]
```diff
@@ -4,16 +4,45 @@ use std::collections::hash_map::{Entry, HashMap};
...
+/// ```
+/// # use rs_ecs::*;
+/// # struct MyResource(u32);
+/// let mut resources = Resources::new();
+///
+/// // Insert multiple resources
+/// resources.insert(42_u32);
+/// resources.insert(MyResource(0));
+///
+/// // Borrow a resource of a type, immutably
```
Comment (adamreichold) [inline src/resources.rs]: I think "of a type" is not helpful as there is no other kind of resource.

File: src/resources.rs [hunk 6]
```diff
@@ -4,16 +4,45 @@ use std::collections::hash_map::{Entry, HashMap};
...
+/// A type map for holding resources.
+///
+/// In an ECS, resources are "global" singular objects that may be accessed by systems
+/// but are not realized as [Entities](struct.Entity.html), as an optimization.
+///
+/// # Examples
+///
+/// ```
+/// # use rs_ecs::*;
+/// # struct MyResource(u32);
```
Comment (adamreichold) [inline src/resources.rs]: Maybe this should not be hidden as declaring wrapper is a common technique?

File: src/resources.rs [hunk 7]
```diff
@@ -26,6 +55,7 @@ impl Resources {
 }
 
 impl Resources {
+    /// Get an immutable reference to a resource.
```
Comment (adamreichold) [inline src/resources.rs]: These panic if the resource is not present.

File: src/resources.rs [hunk 8]
```diff
@@ -49,6 +80,7 @@ impl Resources {
     }
 }
 
+/// An immutable reference to a resource.
```
Comment (adamreichold) [inline src/resources.rs]: Personally, I would prefer "borrow" instead of "reference" for this and the analog `Comp` to emphasise that this type facilitates runtime borrow checking.

File: src/world.rs [hunk 9]
```diff
@@ -47,6 +50,18 @@ fn tag() -> u32 {
 }
 
 impl World {
+    /// Create an [Entity](struct.Entity.html) without any components.
+    /// To add components, see [insert()](struct.World.html#method.insert).
```
Comment (adamreichold) [inline src/world.rs]: I thought there is a simpler syntax to reference crate items in links that does not refer to the resulting HTML files but to the Rust paths?

File: src/world.rs [hunk 10]
```diff
@@ -102,6 +131,21 @@ impl World {
         &self.archetypes
     }
 
+    /// Insert components for a given [Entity](struct.Entity.html).
+    ///
+    /// # Panics
+    ///
+    /// Panics if there is already one of the component types for the entity.
```
Comment (adamreichold) [inline src/world.rs]: I stumbled over this while reading. Maybe "if one of the components is already present for the entity"?

File: src/world.rs [hunk 11]
```diff
@@ -156,6 +213,7 @@ impl World {
 
             self.move_(ent.id, meta.ty, new_ty, meta.idx);
 
+            // TODO: Why use an Option if is always return Some<Bundle>? What happens if a component is not present?
```
Comment (adamreichold) [inline src/world.rs]: There is an early return a few lines above
```rust
B::remove(&mut types)?;
```
which returns `None` if one the components in the given bundle is not present.

File: src/world.rs [hunk 12]
```diff
@@ -201,6 +259,19 @@ impl World {
...
+    /// # Example
+    ///
+    /// ```
+    /// # use rs_ecs::*;
+    /// let mut world = World::new();
+    ///
+    /// let entity = world.alloc();
+    /// world.insert(entity, (42_u32, true));
+    ///
+    /// let has_u32 = world.contains::<u32>(entity);
```
Comment (adamreichold) [inline src/world.rs]: To make the doc tests more powerful, this should actually be an assertion, e.g. `assert!(world.contains::<u32>(entity));`.

File: src/world.rs [hunk 13]
```diff
@@ -211,6 +282,19 @@ impl World {
...
+    /// # Example
+    ///
+    /// ```
+    /// # use rs_ecs::*;
+    /// let mut world = World::new();
+    ///
+    /// let entity = world.alloc();
+    /// world.insert(entity, (42_u32, true));
+    ///
+    /// let comp = world.get::<u32>(entity);
```
Comment (adamreichold) [inline src/world.rs]: `unwrap`ing the component would make this a more powerful test.

File: src/world.rs [hunk 14]
```diff
@@ -221,6 +305,19 @@ impl World {
...
+    /// # Example
+    ///
+    /// ```
+    /// # use rs_ecs::*;
+    /// let mut world = World::new();
+    ///
+    /// let entity = world.alloc();
+    /// world.insert(entity, (42_u32, true));
+    ///
+    /// let comp = world.get_mut::<u32>(entity);
```
Comment (adamreichold) [inline src/world.rs]: `unwrap`ing as a test assertion as above.

File: src/world.rs [hunk 15]
```diff
@@ -236,6 +333,7 @@ impl World {
     }
 }
 
+/// An entity id, with a generation
```
Comment (adamreichold) [inline src/world.rs]: I would probably refrain from making the generation part of the documentation as it is merely an implementation detail to track stale entity handles (and could be solved in other ways). Just an "opaque entity handle" or "opaque entity identifier" seems preferable IMHO.

File: src/resources.rs [hunk 16]
```diff
@@ -4,16 +4,45 @@ use std::collections::hash_map::{Entry, HashMap};
...
+/// resources.insert(MyResource(0));
+///
+/// // Borrow a resource of a type, immutably
+/// let my_res = resources.get::<MyResource>();
+///
+/// // Borrow a resource of a type, mutably
+/// let mut u32_res = resources.get_mut::<u32>();
+/// *u32_res += 1;
+/// ```
 #[derive(Default)]
```
Comment (adamreichold) [inline src/resources.rs]: I wonder if it would actually be lines well spent to manually implement `Default` to be able to document it as you did with `World`?

**Comment by adamreichold** [comment]
As a general courtesy towards crate users, I think there is a lint against undocumented public items which we could deny at crate level? (If that lint is not triggered by those public items which are not actually re-exported that is. I would like to avoid making boiler plate documentation just to satisfy the lint.)

**Comment by adamreichold** [comment]
> > > How can documentation be made foldable, like on docs.rs?
> > 
> > 
> > Could you post a screenshot of what you mean compared to what you get? Additionally, did you trying using `cargo +nightly doc`? I think docs.rs is not always running the stable version of rustdoc.
> 
> At the top right of each doc page, there is `[-][src]`. In the docs for `World`, the `[-]` does not fold method docs. However, I just realized that is does in the docs for `Recources` (commited just now).

Using the latest nightly to build the documentation locally, folding works everywhere for me.

File: src/world.rs (hunk 9 above)
Comment (adamreichold) [reply src/world.rs]: I think I mean [this](https://doc.rust-lang.org/stable/rustdoc/linking-to-items-by-name.html) which was recently stabilised.

File: src/lib.rs (hunk 1 above)
Comment (mlange-42) [reply src/lib.rs]: Not sure if we really need examples. But as I was already doubting, I will remove the long example from the docs.

File: src/lib.rs (hunk 4 above)
Comment (mlange-42) [reply src/lib.rs]: Removed the long example.

File: src/world.rs (hunk 13 above)
Comment (mlange-42) [reply src/world.rs]: Same for `get_mut`

File: src/resources.rs (hunk 8 above)
Comment (mlange-42) [reply src/resources.rs]: Replaced at all occurences

File: src/world.rs (hunk 9 above)
Comment (mlange-42) [reply src/world.rs]: Indeed...

File: src/resources.rs (hunk 16 above)
Comment (mlange-42) [reply src/resources.rs]: Yes, I think this is useful.

**Comment by mlange-42** [comment]
> As a general courtesy towards crate users, I think there is a lint against undocumented public items which we could deny at crate level? (If that lint is not triggered by those public items which are not actually re-exported that is. I would like to avoid making boiler plate documentation just to satisfy the lint.)

Yes, there is `#![warn(missing_docs)]` (or `deny(...)`), and it seems to list only crate-level public items.

Generally, I am not sure whether I can correctly document/describe the more "internal", but public things.

**Comment by adamreichold** [comment]
> > As a general courtesy towards crate users, I think there is a lint against undocumented public items which we could deny at crate level? (If that lint is not triggered by those public items which are not actually re-exported that is. I would like to avoid making boiler plate documentation just to satisfy the lint.)
> 
> Yes, there is `#![warn(missing_docs)]` (or `deny(...)`), and it seems to list only crate-level public items.
> 
> Generally, I am not sure whether I can correctly document/describe the more "internal", but public things.

Personally, I would actually prefer to limit this MR to the effectively public API to keep its scope a small as possible.

If you are feeling generous, I would also be glad if you could squash the fix-ups. I suspect this implies ending up with a single commit but that seems alright to me.

**Comment by adamreichold** [comment]
> Generally, I am not sure whether I can correctly document/describe the more "internal", but public things.

Do you mean the parts of the query API that made the CI fail? If you want to, I could try to push a commit doing those?

File: src/world.rs [hunk 17]
```diff
@@ -6,6 +6,7 @@ use std::sync::atomic::{AtomicU32, Ordering};
 
 use crate::archetype::{Archetype, Comp, CompMut, TypeMetadataSet};
 
+/// The ECS world storing [Entity] and components.
```
Comment (adamreichold) [inline src/world.rs:9]: I don't think Entity should be a link or proper name here, as it is a handle and not the entity itself. Maybe just "entities and their components"?

Comment (adamreichold) [reply src/world.rs:9]: I understand now that you did this consistently and I did warm up to it. So let's ignore my initial reservations...

File: src/world.rs [hunk 18]
```diff
@@ -102,6 +131,21 @@ impl World {
...
+    /// Panics if one of the components is already present for the entity.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # use rs_ecs::*;
+    /// let mut world = World::new();
+    ///
+    /// let entity = world.alloc();
+    /// world.insert(entity, (42_u32, true));
```
Comment (adamreichold) [inline src/world.rs:147]: Maybe add an additional call showing the single-entry-tuple trick?

File: src/world.rs [hunk 19]
```diff
@@ -130,6 +174,19 @@ impl World {
...
+    ///
+    /// world.remove::<(u32, )>(entity);
```
Comment (adamreichold) [inline src/world.rs]: I think Rustfmt would remove that trailing space after the comma? As with insert it is probably helpful to show both single component and component bundle variants in the example? (I understand the two methods together show everything but I think it is nice if each method documents those pitfalls on its own.)

Comment (adamreichold) [reply src/world.rs]: (This would also be a stronger test by `unwrap`ing the result of the call.)

Comment (mlange-42) [reply src/world.rs]: Completely agreed.

**Comment by mlange-42** [comment]
> > Generally, I am not sure whether I can correctly document/describe the more "internal", but public things.
> 
> Do you mean the parts of the query API that made the CI fail? If you want to, I could try to push a commit doing those?

I would continue with `Query`, `With` and `Without`, but certainly have difficulties regarding the remaining types. So, go ahead.

**Comment by adamreichold** [comment]
> > > Generally, I am not sure whether I can correctly document/describe the more "internal", but public things.
> > 
> > 
> > Do you mean the parts of the query API that made the CI fail? If you want to, I could try to push a commit doing those?
> 
> I would continue with `Query`, `With` and `Without`, but certainly have difficulties regarding the remaining types. So, go ahead.

If I am not missing anything, this only leaves out `QuerySpec` and `QueryIter` for I would suggest something along the lines of "type level specification of a query for a certain set of components" and "used to iterate through the entities which match a certain query".

**Comment by mlange-42** [comment]
There are also `Comp` and `CompMut`. But they simply seem equivalent to `Res` and `ResMut`. Surely, `QueryIter` was already clear 😉 

So I think I can do it.

**Comment by adamreichold** [comment]
Concerning the soundness fix API change, I will prepare the necessary changes for swifco-rs but wait with creating a tag for this work?

**Comment by mlange-42** [comment]
> Concerning the soundness fix API change, I will prepare the necessary changes for swifco-rs but wait with creating a tag for this work?

Yes, waiting with the tag sound reasonable. I hope that I can finish this tomorrow.

File: src/query.rs [hunk 20]
```diff
@@ -10,6 +10,26 @@ use crate::{
     world::{Entity, World},
 };
 
+/// Query to get an iterator over entities and components,
```
Comment (adamreichold) [inline src/query.rs]: I think `Entity` can be considered "just another component" in these descriptions? So maybe we could shorten this to something like "Query to get an iterator over all entities with a certain combination of components"?

File: src/query.rs [hunk 21]
```diff
@@ -40,6 +70,20 @@ where
         }
     }
 
+    /// Borrow a query to allow for iteration.
```
Comment (adamreichold) [inline src/query.rs]: I would rather say that this borrows the world, not the query. (I have a commit that I want to push after this PR that renames this lifetime to `'w` because it is how long the world stays borrowed.)

File: src/query.rs [hunk 22]
```diff
@@ -80,13 +124,30 @@ where
         }
     }
 
+    /// Narrow down a query to entities that have a certain component,
+    /// without obtaining that component.
```
Comment (adamreichold) [inline src/query.rs]: Maybe we should strengthen this as the component is not only not fetched, but also not borrowed?

**Comment by mlange-42** [comment]
We should probably also somewhere point to the fact of "prepared queries"...? I guess this is best done in the struct docs?

**Comment by adamreichold** [comment]
> We should probably also somewhere point to the fact of "prepared queries"...? I guess this is best done in the struct docs?

Yes, I think the struct docs of `Query` are the most reasonable place for this information. Ideally we'd touch on why queries are stand-alone (prepared) and hence need to borrow the world before their results can be iterated over.

File: src/query.rs [hunk 23]
```diff
@@ -109,6 +174,7 @@ impl<S> QueryRef<'_, S>
 where
     S: QuerySpec,
 {
+    /// Create an iterator over the query's matches.
```
Comment (adamreichold) [inline src/query.rs]: I think the short possessive form is usually restricted to people. Maybe "Create an iterator over the entities matching the query."?

**Comment by mlange-42** [comment]
> 
> 
> > We should probably also somewhere point to the fact of "prepared queries"...? I guess this is best done in the struct docs?
> 
> Yes, I think the struct docs of `Query` are the most reasonable place for this information. Ideally we'd touch on why queries are stand-alone (prepared) and hence need to borrow the world before their results can be iterated over.

Tried it in the last commit.
